src/sentry/sentry_metrics/querying/visitors/query_expression.py

```diff
@@ -437,3 +460,79 @@ def _visit_float(self, float_number: float) -> QueryExpression:
 
     def _is_numeric_scalar(self, value: QueryExpression) -> bool:
         return isinstance(value, int) or isinstance(value, float)
+
+
+class MapperVisitor(QueryExpressionVisitor):
+    """
+    Visitor that recursively transforms the QueryExpression components to modulate certain attributes to be queried
+    by API that need to be translated for Snuba to be able to query the data.
+    """
+
+    def __init__(self, projects: Sequence[Project], mapper_config: MapperConfig):
+        self.projects = projects
+        self.mapper_config = mapper_config
+        self.mappers: list[Mapper] = []
+
+    def get_or_create_mapper(
+        self, from_key: str | None = None, to_key: str | None = None
+    ) -> Mapper | None:
+        # retrieve the mapper type that is applicable for the given key
+        mapper_class = self.mapper_config.get(from_key=from_key, to_key=to_key)
+        # check if a mapper of the type already exists
+        if mapper_class:
+            for mapper in self.mappers:
+                if mapper_class == type(mapper):
+                    # if a mapper already exists, return the existing mapper
+                    return mapper
+            else:
+                # if no mapper exists yet, instantiate the object and append it to the mappers list
+                mapper_instance = mapper_class()
+                self.mappers.append(mapper_instance)
+                return mapper_instance
+        else:
+            # if no mapper is configured for the key, return None
+            return None
+
+    def _visit_formula(self, formula: Formula) -> Formula:
+        formula = super()._visit_formula(formula)
+        visitor = MapperConditionVisitor(self.projects, self.mapper_config)
+        filters = visitor.visit_group(formula.filters)
+        formula = formula.set_filters(filters)
+        if len(visitor.mappers) > 0:
+            self.mappers.extend(visitor.mappers)
```

Comment:
nit: accessing a field of a class to get the mappers, is assuming that the `visit_group` method is called. To make this more reliable we could return tuple `(mappers, filters)` from `MapperConditionVisitor`, this way we statically guarantee that the mappers can only be returned when you visit a condition tree.